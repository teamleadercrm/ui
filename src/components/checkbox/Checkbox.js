import { IconCheckmarkMediumOutline, IconCheckmarkSmallOutline, IconMinusSmallOutline } from '@teamleader/ui-icons';
import cx from 'classnames';
import omit from 'lodash.omit';
import PropTypes from 'prop-types';
import React, { createRef, PureComponent } from 'react';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';
import theme from './theme.css';

/** @type {React.ComponentType<any>} */
class Checkbox extends PureComponent {
  inputNode = createRef();

  handleToggle = (event) => {
    const { disabled, checked, onChange } = this.props;

    if (!disabled && onChange) {
      onChange(!checked, event);
    }
  };

  render() {
    const { checked, disabled, className, size, label, children, indeterminate, ...others } = this.props;
    const TextElement = size === 'small' ? TextSmall : size === 'medium' ? TextBodyCompact : TextDisplay;
    const IconCheckmark = size === 'large' ? IconCheckmarkMediumOutline : IconCheckmarkSmallOutline;

    const restProps = omit(others, ['onChange']);
    const boxProps = pickBoxProps(restProps);
    const inputProps = omitBoxProps(restProps);

    const classNames = cx(
      theme['checkbox'],
      theme[`is-${size}`],
      {
        [theme['is-checked']]: checked || indeterminate,
        [theme['is-disabled']]: disabled,
      },
      className,
    );

    return (
      <Box element="label" data-teamleader-ui="checkbox" className={classNames} {...boxProps}>
        <input
          className={theme['input']}
          type="checkbox"
          checked={checked}
          disabled={disabled}
          onClick={this.handleToggle}
          ref={this.inputNode}
          readOnly
          {...inputProps}
        />
        <span className={theme['shape']}>
          {checked ? <IconCheckmark className={theme['icon']} /> : <IconMinusSmallOutline className={theme['icon']} />}
        </span>
        {(label || children) && (
          <span className={theme['label']}>
            {label && (
              <TextElement element="span" color={disabled ? 'neutral' : 'teal'}>
                {label}
              </TextElement>
            )}
            {children}
          </span>
        )}
      </Box>
    );
  }
}

Checkbox.propTypes = {
  /** If true, the checkbox will be checked. */
  checked: PropTypes.bool,
  /** The content to display next to the checkbox. */
  children: PropTypes.node,
  /** If true, component will be disabled. */
  disabled: PropTypes.bool,
  /** Name for form input. */
  name: PropTypes.string,
  /** A class name for the wrapper to give custom styles. */
  className: PropTypes.string,
  /** The textual label displayed next to the checkbox. */
  label: PropTypes.string,
  /** Callback function that is fired when checkbox is toggled. */
  onChange: PropTypes.func,
  /** Indicate whether the checkbox is neither checked or unchecked. */
  indeterminate: PropTypes.bool,
  /** Size of the checkbox. */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

Checkbox.defaultProps = {
  checked: false,
  disabled: false,
  indeterminate: false,
  size: 'medium',
};

export default Checkbox;
