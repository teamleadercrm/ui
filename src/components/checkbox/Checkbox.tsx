import { IconCheckmarkMediumOutline, IconCheckmarkSmallOutline, IconMinusSmallOutline } from '@teamleader/ui-icons';
import cx from 'classnames';
import omit from 'lodash.omit';
import React, { ChangeEvent, forwardRef, ReactNode } from 'react';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';
import theme from './theme.css';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';
import { SIZES } from '../../constants';

interface CheckboxProps extends Omit<BoxProps, 'onChange' | 'size'> {
  /** If true, the checkbox will be checked. */
  checked?: boolean;
  /** The content to display next to the checkbox. */
  children?: ReactNode;
  /** If true, component will be disabled. */
  disabled?: boolean;
  /** Name for form input. */
  name?: string;
  /** A class name for the wrapper to give custom styles. */
  className?: string;
  /** The textual label displayed next to the checkbox. */
  label?: string;
  /** Callback function that is fired when checkbox is toggled. */
  onChange?: (checked: boolean, event: ChangeEvent<HTMLInputElement>) => void;
  /** Indicate whether the checkbox is neither checked or unchecked. */
  indeterminate?: boolean;
  /** Size of the checkbox. */
  size?: Exclude<typeof SIZES[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;
}

const Checkbox: GenericComponent<CheckboxProps> = forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {
      checked = false,
      disabled = false,
      className,
      size = 'medium',
      label,
      children,
      indeterminate = false,
      onChange,
      ...others
    },
    ref,
  ) => {
    const handleToggle = (event: ChangeEvent<HTMLInputElement>) => {
      if (onChange) {
        onChange(event.target.checked, event);
      }
    };

    const TextElement = size === 'small' ? TextSmall : size === 'medium' ? TextBodyCompact : TextDisplay;
    const IconCheckmark = size === 'large' ? IconCheckmarkMediumOutline : IconCheckmarkSmallOutline;

    const restProps = omit(others, ['onChange']);
    const boxProps = pickBoxProps(restProps);
    const inputProps = omitBoxProps(restProps);

    const classNames = cx(
      theme['checkbox'],
      theme[`is-${size}`],
      {
        [theme['is-checked']]: checked || indeterminate,
        [theme['is-disabled']]: disabled,
      },
      className,
    );

    return (
      <Box element="label" data-teamleader-ui="checkbox" className={classNames} {...boxProps}>
        <input
          className={theme['input']}
          type="checkbox"
          checked={checked}
          disabled={disabled}
          onChange={handleToggle}
          ref={ref}
          readOnly
          {...inputProps}
        />
        <span className={theme['shape']}>
          {checked ? <IconCheckmark className={theme['icon']} /> : <IconMinusSmallOutline className={theme['icon']} />}
        </span>
        {(label || children) && (
          <span className={theme['label']}>
            {label && (
              <TextElement element="span" color={disabled ? 'neutral' : 'teal'}>
                {label}
              </TextElement>
            )}
            {children}
          </span>
        )}
      </Box>
    );
  },
);

Checkbox.displayName = 'Checkbox';

export default Checkbox;
