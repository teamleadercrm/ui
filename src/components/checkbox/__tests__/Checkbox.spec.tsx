import React from 'react';
import { render } from '@testing-library/react';
import Checkbox from '../Checkbox';

describe('Component - Checkbox', () => {
  it('renders a checked checkbox', async () => {
    const { asFragment } = render(<Checkbox checked>Foobar</Checkbox>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an unchecked checkbox', async () => {
    const { asFragment } = render(<Checkbox>Foobar</Checkbox>);
    expect(asFragment()).toMatchSnapshot();
  });
});
