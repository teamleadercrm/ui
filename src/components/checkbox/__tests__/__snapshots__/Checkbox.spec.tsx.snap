// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Checkbox renders a checked checkbox 1`] = `
<DocumentFragment>
  <label
    class="box checkbox is-medium is-checked"
    data-teamleader-ui="checkbox"
  >
    <input
      checked=""
      class="input"
      readonly=""
      type="checkbox"
    />
    <span
      class="shape"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="14"
        preserveAspectRatio="xMidYMid meet"
        style="opacity: 0.84; vertical-align: middle;"
        viewBox="0 0 14 14"
        width="14"
      >
        <path
          clip-rule="evenodd"
          d="M5.857 10.556L2.293 6.993a.999.999 0 111.414-1.415l2.15 2.15 4.436-4.435a1 1 0 011.414 1.413l-5.85 5.85z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="label"
    >
      Foobar
    </span>
  </label>
</DocumentFragment>
`;

exports[`Component - Checkbox renders an unchecked checkbox 1`] = `
<DocumentFragment>
  <label
    class="box checkbox is-medium"
    data-teamleader-ui="checkbox"
  >
    <input
      class="input"
      readonly=""
      type="checkbox"
    />
    <span
      class="shape"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="14"
        preserveAspectRatio="xMidYMid meet"
        style="opacity: 0.84; vertical-align: middle;"
        viewBox="0 0 14 14"
        width="14"
      >
        <path
          clip-rule="evenodd"
          d="M12 7a1 1 0 01-1 1H3a1 1 0 010-2h8a1 1 0 011 1z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="label"
    >
      Foobar
    </span>
  </label>
</DocumentFragment>
`;
